jaxopt/_src/levenberg_marquardt.py

```diff
@@ -0,0 +1,356 @@
+                                      dparams_diff,
+                                      small_dparams_test_init_state, *args,
+                                      **kwargs):
+
+    def small_dparams_test_pred(dparams_diff):
+      return dparams_diff <= 0.0
+
+    def small_dparams_test_true_func(params, error, grad_error, damping_factor,
+                                     increase_factor, gradient, residual):
+      return (params, 0.0, grad_error, damping_factor, increase_factor,
+              gradient, residual)
+
+    def small_dparams_test_false_func(params, error, grad_error, damping_factor,
+                                      increase_factor, gradient, residual):
+
+      updated_params = params + delta_params
+
+      residual_next = self._fun(updated_params, *args, **kwargs)
+
+      # Calculate denominator of the gain ratio based on Eq. 6.16, "Introduction
+      # to optimization and data fitting", L(0)-L(hlm)=0.5*hlm^T*(mu*hlm-g).
+      gain_ratio_denom = 0.5 * jnp.dot(
+          jnp.transpose(delta_params),
+          (damping_factor * delta_params - gradient))
+
+      # Current value of loss function F=0.5*||f||^2.
+      loss_next = 0.5 * jnp.dot(residual_next, residual_next)
+
+      gain_ratio = (loss_curr - loss_next) / gain_ratio_denom
+
+      gain_ratio_test_init_state = (params, error, grad_error, damping_factor,
+                                    increase_factor, residual, gradient,
+                                    updated_params)
+
+      # Calling the jax condition function:
+      # Note that only the parameters that are used in the rest of the program
+      # are returned by this function.
+
+      params, error, grad_error, damping_factor, increase_factor, gradient, residual = (
+          self.update_state_using_gain_ratio(gain_ratio,
+                                             gain_ratio_test_init_state, *args,
+                                             **kwargs))
+
+      return params, error, grad_error, damping_factor, increase_factor, gradient, residual
+
+    params, error, grad_error, damping_factor, increase_factor, gradient, residual = (
+        jax.lax.cond(
+            small_dparams_test_pred(dparams_diff), small_dparams_test_true_func,
+            small_dparams_test_false_func, *small_dparams_test_init_state))
+
+    return params, error, grad_error, damping_factor, increase_factor, gradient, residual
+
+  def update(self, params, state: NamedTuple, *args, **kwargs) -> base.OptStep:
+    """Performs one iteration of the least-squares solver.
+
+    Args:
+      params: pytree containing the parameters.
+      state: named tuple containing the solver state.
+
+    Returns:
+      (params, state)
+    """
+
+    # Current value of the loss function F=0.5*||f||^2.
+    loss_curr = state.loss
+
+    def jtj_reg_matvec(vec):
+      """The (J^T.J + damping_factor*I) matrix vector product operator."""
+      return self._jtj_reg_op(params, state.damping_factor, vec, *args,
```

Comment:
Since you added the `ridge` option directly to the solvers I think this should be
```suggestion
      return self._jtj_op(params, state.damping_factor, vec, *args,
```
And in this case `def _jtj_reg_op` can be deleted.

Reply:
Actually `jtj_reg_matvec ` can be deleted now, since all of the solvers are using `ridge` and I kept the `jtj_reg_matvec` for the custom solver. If we are assuming custom solver will have the `ridge` option, we can just remove this definition.